File: core/java/src/main/java/hara/lang/protocol/IClosed.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Observable terminal-state capability. */
@HaraProtocolBinding(
    namespace = "std.protocol.iclosed",
    name = "IClosed",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IClosed {
  @HaraMethod(value = "closed?", arity = 1)
  boolean closed();
}

File: core/rust/src/lang/protocol/iclosed.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iclosed",
    name = "IClosed",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IClosed {
    #[hara_method(value = "closed?", arity = 1)]
    fn closed(&self) -> bool;
}
